docs/usage/data-modeling.md

```diff
@@ -0,0 +1,150 @@
+  perspective this is a liability.
+- Alice and Bob cannot "overdraw" their account — that is, their balance will never be negative.
+
+## `user_data`
+
+`user_data` is the most flexible field in the schema (for both
+[accounts](../reference/accounts.md#user_data) and [transfers](../reference/transfers.md#user_data)).
+`user_data`'s contents are arbitrary, interpreted only by the application.
+
+`user_data` is indexed for efficient point and range queries.
+
+Example uses:
+
+- Set `user_data` to a "foreign key" — that is, an identifier of a corresponding object within
+  another database.
+- Set `user_data` to a group identifier for objects that will be queried together.
+- Set `user_data` to a transfer or account `id`.
+  (TODO: Can we use this for join queries via the query API, or must the application implement them?)
+
+## `id`
+
+The primary purpose of an `id` (for both [accounts](../reference/accounts.md#id) and
+[transfers](../reference/transfers.md#id)) is to serve as an "idempotency key".
+The database uses ids to avoid repeating duplicate events.
+
+[Randomly-generated identifiers](#random-identifier) are recommended for most applications.
+
+When selecting an `id` scheme:
+
+- Idempotency is particularly important (and difficult) in the context of
+  [application crash recovery](./integration.md#consistency-with-foreign-databases).
+- Be careful to [avoid `id` collisions](https://en.wikipedia.org/wiki/Birthday_problem).
+- An account and a transfer may share the same `id` — they belong to different "namespaces".
+- Avoid requiring a central oracle to generate each unique `id` (e.g. an auto-increment field in SQL).
+  A central oracle may become a performance bottleneck when creating accounts/transfers.
+- `id` is mostly accessed by point queries, but it is indexed for efficient range queries as well.
+
+### Examples
+#### Random Identifier
+
+Randomly-generated identifiers are recommended for most applications.
+
+- Random identifiers require coordination with a secondary database to implement idempotent
+  [application crash recovery](./integration.md#consistency-with-foreign-databases).
+- Random identifiers have an insignificant risk of collisions.
+- Random identifiers do not require a central oracle.
+- Only point queries are useful for fetching randomly-generated identifiers.
+
+To maximize id entropy, prefer a cryptographically-secure PRNG.
+
+#### Reuse Foreign Identifier
+
+This technique is most appropriate when integrating TigerBeetle with an existing application
+where TigerBeetle accounts or transfers map one-to-one with an entity in the foreign database.
+
+Set `id` to a "foreign key" — that is, reuse an identifier of a corresponding object from another
+database. For example, if every user (within the application's database) has a single account, then
+the identifier within the foreign database can be used as the `Account.id` within TigerBeetle.
```

Comment:
Should we include a warning about bigserial account ids and transfer ids or is that too obscure?

Reply:
I'm not familiar with bigserial ids, but if it is database-specific it probably doesn't belong here. (I am curious what you mean though).

Reply:
ah sorry, autoincrement : mysql, bigserial: postgres

Reply:
Ah, gotcha. I don't think we need to mention it. If they are already using that in their 'foreign database' then it is actually ok, it just isn't what we recommend when starting from scratch.

Reply:
well it won't be ok if we intend to collapse namespaces. but maybe we are still sticking to strategic silence in docs here for the moment while we decide :D

Reply:
Oh, I forgot about that. I'll leave it for now & change it if we combine the namespaces.